presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusRecordCursor.java

```diff
@@ -0,0 +1,302 @@
+import com.facebook.presto.common.block.BlockBuilder;
+import com.facebook.presto.common.type.ArrayType;
+import com.facebook.presto.common.type.MapType;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.common.type.TypeUtils;
+import com.facebook.presto.common.type.VarcharType;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.RecordCursor;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.io.ByteSource;
+import com.google.common.io.CountingInputStream;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.common.type.DateTimeEncoding.packDateTimeWithZone;
+import static com.facebook.presto.common.type.DoubleType.DOUBLE;
+import static com.facebook.presto.common.type.StandardTypes.BIGINT;
+import static com.facebook.presto.common.type.StandardTypes.BOOLEAN;
+import static com.facebook.presto.common.type.StandardTypes.INTEGER;
+import static com.facebook.presto.common.type.StandardTypes.SMALLINT;
+import static com.facebook.presto.common.type.StandardTypes.TINYINT;
+import static com.facebook.presto.common.type.VarcharType.createUnboundedVarcharType;
+import static com.facebook.presto.plugin.prometheus.PrometheusClient.TIMESTAMP_COLUMN_TYPE;
+import static com.facebook.presto.spi.StandardErrorCode.NOT_SUPPORTED;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class PrometheusRecordCursor
+        implements RecordCursor
+{
+    private final List<PrometheusColumnHandle> columnHandles;
+    private final int[] fieldToColumnIndex;
+
+    private final Iterator<PrometheusStandardizedRow> metricsItr;
+    private final long totalBytes;
+
+    private PrometheusStandardizedRow fields;
+
+    public PrometheusRecordCursor(List<PrometheusColumnHandle> columnHandles, ByteSource byteSource)
+    {
+        this.columnHandles = columnHandles;
+
+        fieldToColumnIndex = new int[columnHandles.size()];
+        for (int i = 0; i < columnHandles.size(); i++) {
+            PrometheusColumnHandle columnHandle = columnHandles.get(i);
+            fieldToColumnIndex[i] = columnHandle.getOrdinalPosition();
+        }
+
+        try (CountingInputStream input = new CountingInputStream(byteSource.openStream())) {
+            metricsItr = prometheusResultsInStandardizedForm(new PrometheusQueryResponse(input).getResults()).iterator();
+            totalBytes = input.getCount();
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
```

Comment:
shall we define an error code, and add some information to the exception?

Reply:
sure, I added 2 catches, one from parsing of`PrometheusQueryResponse`, one for mapping the result